lib/internal/streams/operators.js

```diff
@@ -0,0 +1,115 @@
+'use strict';
+
+const console = require('console');
+const { AbortController } = require('internal/abort_controller');
+const { AbortError } = require('internal/errors');
+const Readable = require('internal/streams/readable');
+const eos = require('internal/streams/end-of-stream');
+
+module.exports.map = function map(stream, fn, options) {
+  // TODO: Argument validation
+
+  let concurrency = 1;
+  if (Number.isFinite(options)) {
+    concurrency = options;
+  } else if (options && Number.isFinite(options.concurrency)) {
+    concurrency = options.concurrency;
+  }
+
+  let highWaterMark = 1;
+  if (options && Number.isFinite(options.highWaterMark)) {
```

Comment:
Why are we supporting _both_ highWaterMark _and_ concurrency?

I get that conceptually it's kind of different (highWaterMark is when we stop asking for more data from the source and concurrency is how much data we're asking for in parallel) but I'm wondering if this API isn't a bit too confusing?

Reply:
It's two different things:

- concurrency: how many active calls to `fn`
- highWaterMark: how many items do we have in buffer

e.g. maybe I'd like to buffer up to 128 elements but I only want to have 4 http requests active at a time.

I'm fine with removing it.

Reply:
Yeah, I understand the rationale but I am not sure people will get the nuance and I'm a bit afraid of giving them this.